tiny_dnn/layers/global_average_pooling_layer.h

```diff
@@ -0,0 +1,130 @@
+/*
+    Copyright (c) 2015, Taiga Nomi
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+
+#include <algorithm>
+#include <string>
+#include <vector>
+
+#include "tiny_dnn/layers/layer.h"
+#include "tiny_dnn/util/util.h"
+
+#include "tiny_dnn/core/kernels/global_avepool_grad_op.h"
+#include "tiny_dnn/core/kernels/global_avepool_op.h"
+
+namespace tiny_dnn {
+
+/**
+ * applies channel-wise global average pooling to spatial data.
+ **/
+class global_average_pooling_layer : public layer {
+ public:
+  using layer::parallelize_;
+
+  global_average_pooling_layer(const shape3d &in_shape,
+                               backend_t backend_type = core::default_engine())
+    : global_average_pooling_layer(
+        in_shape.width_, in_shape.height_, in_shape.depth_, backend_type) {}
+
+  /**
+   * @param in_width     [in] width of input image
+   * @param in_height    [in] height of input image
+   * @param in_channels  [in] the number of input image channels (depth)
+  **/
+  global_average_pooling_layer(serial_size_t in_width,
+                               serial_size_t in_height,
+                               serial_size_t in_channels,
+                               backend_t backend_type = core::default_engine())
+    : layer({vector_type::data}, {vector_type::data}) {
+    set_global_avepool_params(shape3d(in_width, in_height, in_channels),
+                              shape3d(in_channels, 1, 1));
+
+    init_backend(backend_type);
+    layer::set_backend_type(backend_type);
```

Comment:
it makes more sense to set the backend type inside the `init_backend`routine. Others layers will also need this -> next PR